frontend/lib/resolution/scope-queries.cpp

```diff
@@ -840,11 +843,33 @@ bool LookupHelper::doLookupInScope(const Scope* scope,
     traceCurPathSize = traceCurPath->size();
   }
 
-  auto pair = checkedScopes.insert(std::make_pair(name, scope));
-  if (pair.second == false) {
-    // scope has already been visited by this function,
-    // so don't try it again.
-    return false;
+  // update the checkedScopes map and return early if there is nothing to do.
+  auto p = checkedScopes.insert(std::make_pair(CheckedScope(name, scope),
+                                curFilter));
+  if (p.second == false) {
+    // the insert did not succeed: there was already something in the map.
+    // decide what to do about it.
+    IdAndVis::SymbolTypeFlags foundFilter = p.first->second;
+    if ((curFilter & foundFilter) == foundFilter) {
+      // if the flags we found are equal to foundFilter,
+      // or if curFilter is a superset of foundFilter
+      // (which, because these are filters, means that foundFilter is
+      //  less restricted / more general),
+      // there is no need to visit this scope further.
+      return false;
+    }
+
+    // otherwise, compute the new filter to use
+    IdAndVis::SymbolTypeFlags onlyInFound = 0;
+    onlyInFound = foundFilter & ~curFilter;
+    curFilter = IdAndVis::reverseFlags(onlyInFound);
+
+    // update checkedScopes to remove filter bits that weren't present
+    // in foundFilter (because we are going to update results
+    // with matches for the now-not-filtered-out cases)
+    IdAndVis::SymbolTypeFlags combinedFilter = 0;
+    combinedFilter = foundFilter & ~onlyInFound;
```

Comment:
This works out to `foundFilter & originalCurFilter`, which I think might be more clear.